### PR TITLE
Add Enderman Mock

### PR DESCRIPTION
**Pull request by thelooter**
# Description
Add Enderman Mock

# Checklist
The following items should be checked before the pull request can be merged.
- [x] Unit tests added.
- [x] Code follows existing code format.

# Info on creating a pull request
- Make sure that unit tests are added which test the relevant changes.
- Make sure that the changes follow the existing code format.

# For maintainer
When a PR is approved, the maintainer should label this PR with `release/*`.

- If the PR fixes a bug in MockBukkit, update the patch version. (if the current version is `0.4.1`, the new version should be `0.4.2`)
- If the PR adds a new feature to MockBukkit, update minor version. (if the current version is `0.4.1`, the new version should be `0.5.0`)

Note that a PR that fixes an `UnimplementedOperationException` should be considered a new version and not a bugfix.

### CODE REVIEW

File: src/main/java/be/seeseemelk/mockbukkit/WorldMock.java [hunk 1]
```diff
@@ -880,6 +882,9 @@ else if (clazz == Zombie.class)
 		{
 			return new ZombieMock(server, UUID.randomUUID());
 		}
+		else if (clazz == Enderman.class)
+		{
+			return new EndermanMock(server, UUID.randomUUID());
 		else if (clazz == Sheep.class)
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/WorldMock.java]:
```suggestion
			return new EndermanMock(server, UUID.randomUUID());
		}
		else if (clazz == Sheep.class)
```

File: src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+import java.util.UUID;
+
+public class EndermanMock extends MonsterMock implements Enderman
+{
+
+	private MaterialData carriedMaterial = null;
+	private BlockData carriedBlock = null;
+	boolean isScreaming = false;
+	boolean hasBeenStaredAt = false;
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java]:
```suggestion
	private boolean isScreaming = false;
	private boolean hasBeenStaredAt = false;
```

File: src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+	public EndermanMock(ServerMock server, UUID uuid)
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java]:
```suggestion
	public EndermanMock(@NotNull ServerMock server, @NotNull UUID uuid)
```

File: src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	{
+		// TODO Auto-generated method stub
+		throw new UnimplementedOperationException();
+	}
+
+	@Override
+	public @NotNull MaterialData getCarriedMaterial()
+	{
+		if (carriedMaterial == null){
+			throw  new IllegalStateException("Carried Material must be set before using this method");
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java]:
```suggestion
		if (carriedMaterial == null)
		{
			throw new IllegalStateException("Carried Material must be set before using this method");
```

File: src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	{
+		Preconditions.checkNotNull(material, "MaterialData cannot be null");
+		this.carriedMaterial = material;
+	}
+
+	@Override
+	public @Nullable BlockData getCarriedBlock()
+	{
+		if (carriedBlock == null){
+			throw  new IllegalStateException("Carried Block must be set before using this method");
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java]:
```suggestion
		if (carriedBlock == null)
		{
			throw new IllegalStateException("Carried Block must be set before using this method");
```

**Review by Insprill — CHANGES_REQUESTED** [review]

File: src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.UUID;
+
+public class EndermanMock extends MonsterMock implements Enderman
+{
+
+	private MaterialData carriedMaterial = null;
+	private BlockData carriedBlock = null;
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java]: There shouldn't be two separate fields for material and block. There should only be one BlockState that the material methods get the type of, and set to from `BlockStateMock#mock`

**Review by Insprill — CHANGES_REQUESTED** [review]

File: src/test/java/be/seeseemelk/mockbukkit/entity/EndermanMockTest.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.UUID;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class EndermanMockTest
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/entity/EndermanMockTest.java]:
```suggestion
class EndermanMockTest
```

**Review by Insprill — CHANGES_REQUESTED** [review]

File: src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+		throw new UnimplementedOperationException();
+	}
+
+	@Override
+	public @NotNull MaterialData getCarriedMaterial()
+	{
+		if (carriedBlock == null)
+		{
+			throw new IllegalStateException("Carried Block must be set before using this method");
+		}
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java]:
```suggestion
		checkHasBlock();
```

File: src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+		return this.hasBeenStaredAt;
+	}
+
+	@Override
+	public void setHasBeenStaredAt(boolean hasBeenStaredAt)
+	{
+		this.hasBeenStaredAt = hasBeenStaredAt;
+	}
+
+}
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java:91]:
```suggestion
	}

	private void checkHasBlock()
	{
		Preconditions.checkState(carriedBlock != null, "Carried Block must be set before using this method");
	}

}
```

File: src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+		carriedBlock = BlockDataMock.mock(material.getItemType());
+	}
+
+	@Override
+	public @Nullable BlockData getCarriedBlock()
+	{
+		if (carriedBlock == null)
+		{
+			throw new IllegalStateException("Carried Block must be set before using this method");
+		}
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EndermanMock.java]:
```suggestion
		checkHasBlock();
```

File: src/test/java/be/seeseemelk/mockbukkit/WorldMockTest.java [hunk 11]
```diff
@@ -937,4 +938,13 @@ void testCallSpawnEventOnDisallowedAnimal()
 		assertTrue(pig.isDead());
 	}
 
+	@Test
+	void testSpawnEnderMan()
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/WorldMockTest.java]:
```suggestion
	void testSpawnEnderman()
```

File: src/test/java/be/seeseemelk/mockbukkit/entity/EndermanMockTest.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+		serverMock = MockBukkit.mock();
+		endermanMock = new EndermanMock(serverMock, UUID.randomUUID());
+	}
+
+	@AfterEach
+	void tearDown()
+	{
+		MockBukkit.unmock();
+	}
+
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/entity/EndermanMockTest.java:36]: Tests should be added to ensure the IllegalStateException is thrown when the enderman doesn't have a block.

**Review by Insprill — CHANGES_REQUESTED** [review]

File: src/test/java/be/seeseemelk/mockbukkit/entity/EndermanMockTest.java (hunk 12 above)
Comment (thelooter) [reply src/test/java/be/seeseemelk/mockbukkit/entity/EndermanMockTest.java:36]: There are 2  Tests for this

**Review by Insprill — APPROVED** [review]
LGTM

**Pull request merged**